src/Workspaces/CSharp/Portable/Recommendations/CSharpRecommendationServiceRunner.cs

```diff
@@ -205,9 +206,55 @@ private ImmutableArray<ISymbol> GetSymbolsForTypeOrNamespaceContext()
                 return symbols.WhereAsArray(s => !s.IsDelegateType() && !s.IsInterfaceType());
             }
 
+            if (_context.IsAsyncMemberDeclarationContext)
+            {
+                return symbols.WhereAsArray(IsSymbolValidForAsyncDeclarationContext);
+            }
+
             return symbols;
         }
 
+        private bool IsSymbolValidForAsyncDeclarationContext(ISymbol symbol)
+        {
+            if (symbol.IsNamespace())
+            {
+                return true;
+            }
+
+            if (symbol is not INamedTypeSymbol namedType ||
+                symbol.IsDelegateType() ||
+                namedType.IsEnumType())
+            {
+                return false;
+            }
+
+            if (namedType.TypeKind == TypeKind.Interface)
+            {
+                // The only interfaces, that are valid in async context are IAsyncEnumerable and IAsyncEnumerator.
+                // So if we are validating an interface, then we can just check for 2 of this possible variants
+                return (namedType.Name == "IAsyncEnumerable" || namedType.Name == "IAsyncEnumerator") &&
+                    namedType.TypeParameters.Length == 1 &&
+                    namedType.IsFromSystemRuntimeOrMscorlibAssembly();
```

Comment:
why is this important?  if someone has types like that themselves, why would it be bad to show them?  aren't they likely to be something that can be on async.

Reply:
`IAsyncEnumerable` and `IAsyncEnumerator` can probably not be checked for a containing assembly. For 99.999% of cases just checking by name would be fine. However, `Task` seems to be a lot more common name, so it is wise IMO to check it for a containing assembly. User can develop a system of remote education, and `Task` can represent a task from mentor to a student for instance. But if I have to implement logic of assembly checking anyway, why not to apply it to other cases as well?